include/rmm/mr/device/arena_memory_resource.hpp

```diff
@@ -208,22 +209,24 @@ class arena_memory_resource final : public device_memory_resource {
    *
    * @return std::shared_ptr<arena> The arena associated with the current thread
    */
-  std::shared_ptr<arena> get_arena()
+  arena& get_arena()
   {
-    thread_local auto a = std::make_shared<arena>(upstream_mr_);
-    thread_local bool is_initialized{false};
-
-    if (!is_initialized) {
-      lock_guard lock(mtx_);
-      arenas_.emplace(std::this_thread::get_id(), a);
-      is_initialized = true;
+    auto id = std::this_thread::get_id();
```

Comment:
I liked the `thread_local` solution better. This solution is problematic because thread ids can and will be recycled, so if you have threads being spawned/joined throughout the liftetime of the arena allocator, you could end up with two different threads using the same `arena` which seems problematic.

Reply:
How would you handle the lifetime of the shared pointer with thread_local? That seems to be the cause for the segfault in the test.

Reusing an arena should be ok, as long as the old thread with the same id is no longer active.

Reply:
I assumed that since the `unordered_map` of arenas stored the arena as a `shared_ptr` it should keep the thread's `arena` object alive even if the thread was destroyed and the `thread_local` copied of the `area` was destroyed. Though I'm realizing now that you would still run into the problem of a thread_id being recycled.

Reply:
I think the problem is the other around. When the `arena_memory_resource` is deleted, the thread_local shared_ptr would hold on to an invalid pointer.

Reply:
Are you 100% sure reusing arenas across threads is safe? Especially in PTDS mode where two threads can have the same `id`, but different `cudaStream_t`s.

Reply:
According to this (https://man7.org/linux/man-pages/man3/pthread_self.3.html):
>Thread IDs are guaranteed to be unique only within a process.  A thread ID may be reused after a terminated thread has been joined, or a detached thread has terminated.

So I assume we won't have two *live* threads with the same id. If one thread dies, another thread reuses the id, it should be ok to reuse the arena as well.

Reply:
Yeah, I'm not suggesting two live threads would share an id. I was talking about an id being recycled. 

That leads me to another question. How do you clean up `arenas` when a thread dies? If you have many threads that are created/destroyed using this resource, then you'll potentially end up with many unused `arena`s.

Reply:
This is not a problem for Spark jobs as the number of threads is tied to the number of cpu cores. To make this more robust for other use cases, we can probably add a background thread to do cleanup.

Reply:
I'd like to avoid the complexity of a background thread, but I would like to make our resources generally useful. We need to support non-spark use cases.  I would think this could be done with a thread_local arena. I didn't see the original solution before you changed it.

Reply:
Thought about this more, actually I don't think we need to do additional cleanups. Let's say we have a producer thread that allocates a bunch of memory to hand off to other threads. When we join with the producer thread, we need to keep the corresponding arena around so that we can free the allocations. As the consumer threads free each allocation, the now orphaned arena would become emptier and emptier. If the thread id gets reused and the orphaned arena is put back into action, any remaining free blocks can be reused.

Reply:
OK, but what if it the thread ID is never reused? The orphaned arena has a free list of blocks taking space. Is there a mechanism for other arenas to use those blocks?

Reply:
Added an arena cleaner for thread early termination. Please take another look.